### PR TITLE
feat: New Feature Document Scanner 

### PR DESCRIPTION
**Pull request by bensonarafat**
Document Scanner feature based the new feature request  #581 

Key point to note: 
- Low binary size impact (all ML models and large resources are downloaded centrally in Google Play services). 
-  No camera permission is required - the document scanner leverages the Google Play services' camera permission, and users are in control of which files to share back with your app. 

Document Scanner is currently in Beta, and not currently available for iOS (I will stay updated and update this feature when available)
Read more on the [Google ML doc](https://developers.google.com/ml-kit/vision/doc-scanner )

Thanks [ablause](https://github.com/ablause) on the `onAttachedToActivity` issue 

## Test Sample 

https://github.com/flutter-ml/google_ml_kit_flutter/assets/12992332/5985649b-857e-4bd4-9f03-b058418b5674

### CODE REVIEW

**Comment by patassel** [comment]
Great result ! Good job guys

**Comment by fbernaly** [comment]
I will review, after that I will merge and release a new version. I the meantime use @bensonarafat fork if you need the feature.

File: packages/google_mlkit_document_scanner/android/build.gradle [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+               outputs.upToDateWhen {false}
+               showStandardStreams = true
+            }
+        }
+    }
+}
+
+dependencies {
+   implementation 'com.google.android.gms:play-services-mlkit-document-scanner:16.0.0-beta1'
+//   implementation files('/Users/bensonarafat/development/flutter/bin/cache/artifacts/engine/android-arm64-release/flutter.jar')
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/android/build.gradle:58]: delete this comment

File: packages/google_mlkit_document_scanner/android/build.text [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            }
+        }
+    }
+    buildToolsVersion '30.0.3'
+    ndkVersion '26.2.11394342'
+}
+
+dependencies {
+   implementation 'com.google.android.gms:play-services-mlkit-document-scanner:16.0.0-beta1'
+//    implementation files('/Users/bensonarafat/development/flutter/bin/cache/artifacts/engine/android-arm64-release/flutter.jar')
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/android/build.text:60]: delete this comment

File: packages/google_mlkit_document_scanner/ios/google_mlkit_document_scanner.podspec [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#
+# To learn more about a Podspec see http://guides.cocoapods.org/syntax/podspec.html.
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/ios/google_mlkit_document_scanner.podspec:2]: make this podspec file similar to the other podspecs

File: packages/google_mlkit_document_scanner/pubspec.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+
+flutter:
+  plugin:
+    platforms:
+      android:
+        package: com.google_mlkit_document_scanner
+        pluginClass: GoogleMlkitDocumentScannerPlugin
+      ios:
+        pluginClass: GoogleMlkitDocumentScannerPlugin
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/pubspec.yaml:28]: add extra line add the end

**Comment by fbernaly** [comment]
use path to pass it, and also make sure the job is added to this file:

https://github.com/flutter-ml/google_ml_kit_flutter/blob/develop/.github/workflows/flutter.yml

they are arrange in alphabetical order.

**Comment by bensonarafat** [comment]
> use path to pass it, and also make sure the job is added to this file:
> 
> https://github.com/flutter-ml/google_ml_kit_flutter/blob/develop/.github/workflows/flutter.yml
> 
> they are arrange in alphabetical order.

Yeah, i finally got it. 
Thanks @fbernaly

File: packages/google_ml_kit/pubspec.yaml [hunk 5]
```diff
@@ -2,6 +2,7 @@ name: google_ml_kit
 description: A Flutter plugin to use all APIs from Google's standalone ML Kit for mobile platforms.
 version: 0.16.3
 homepage: https://github.com/flutter-ml/google_ml_kit_flutter
+publish_to: "none"
```
Comment (fbernaly) [inline packages/google_ml_kit/pubspec.yaml:5]: discard changes for this file. We are going to deprecate this umbrella plugin.

File: packages/google_mlkit_document_scanner/README.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+- Since the plugin uses platform channels, you may encounter issues with the native API. Before submitting a new issue, identify the source of the issue. You can run both iOS and/or Android native [example apps by Google](https://github.com/googlesamples/mlkit) and make sure that the issue is not reproducible with their native examples. If you can reproduce the issue in their apps then report the issue to Google. The [authors](https://github.com/flutter-ml/google_ml_kit_flutter/blob/master/AUTHORS) do not have access to the source code of their native APIs, so you need to report the issue to them. If you find that their example apps are okay and still you have an issue using this plugin, then look at our [closed and open issues](https://github.com/flutter-ml/google_ml_kit_flutter/issues). If you cannot find anything that can help you then report the issue and provide enough details. Be patient, someone from the community will eventually help you.
+
+## Requirements
+
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/README.md:53]: Add section for iOS and mentioned that it is not supported yet.

File: packages/google_mlkit_document_scanner/android/build.gradle [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+apply plugin: 'com.android.library'
+
+android {
+    if (project.android.hasProperty("namespace")) {
+        namespace 'com.google_mlkit_document_scanner'
+    }
+
+    compileSdkVersion 33
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/android/build.gradle:29]: All other `build.gradle` files for the other plugins use :
`compileSdk 34`
Change it to match that as well.

File: packages/google_mlkit_document_scanner/android/build.gradle [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_1_8
+        targetCompatibility JavaVersion.VERSION_1_8
+    }
+
+    defaultConfig {
+        minSdkVersion 21
+    }
+
+    dependencies {
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/android/build.gradle:40]: remove L40-L53
Make sure this build.gradle is similar to the other plugins.

File: packages/google_mlkit_document_scanner/android/build.gradle [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+               outputs.upToDateWhen {false}
+               showStandardStreams = true
+            }
+        }
+    }
+}
+
+dependencies {
+   implementation 'com.google.android.gms:play-services-mlkit-document-scanner:16.0.0-beta1'
+}
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/android/build.gradle:58]: Add extra line at the end.

File: packages/google_mlkit_document_scanner/android/build.text [hunk 10]
```diff
@@ -0,0 +1,60 @@
+group 'com.google_mlkit_document_scanner'
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/android/build.text:1]: remove this file

File: packages/google_mlkit_document_scanner/android/gradle/wrapper/gradle-wrapper.properties [hunk 11]
```diff
@@ -0,0 +1,7 @@
+distributionBase=GRADLE_USER_HOME
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/android/gradle/wrapper/gradle-wrapper.properties:1]: we do not have a similar file for the other plugins. Remove this file.

File: packages/google_mlkit_document_scanner/pubspec.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  flutter_test:
+    sdk: flutter
+  flutter_lints: ^2.0.2
+
+flutter:
+  plugin:
+    platforms:
+      android:
+        package: com.google_mlkit_document_scanner
+        pluginClass: GoogleMlkitDocumentScannerPlugin
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/pubspec.yaml:24]: Name should be `GoogleMlKitDocumentScannerPlugin`

File: packages/google_mlkit_document_scanner/android/src/main/java/com/google_mlkit_document_scanner/GoogleMlkitDocumentScannerPlugin.java [hunk 13]
```diff
@@ -0,0 +1,238 @@
+package com.google_mlkit_document_scanner;
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/android/src/main/java/com/google_mlkit_document_scanner/GoogleMlkitDocumentScannerPlugin.java:1]: Follow the same design we have for the other plugins. You should have 2 files: `DocumentScanner.java` and `GoogleMlKitDocumentScannerPlugin.java`

**Comment by fbernaly** [comment]
@bensonarafat : I have more comments.

File: packages/google_mlkit_document_scanner/android/src/main/java/com/google_mlkit_document_scanner/GoogleMlkitDocumentScannerPlugin.java (hunk 13 above)
Comment (bensonarafat) [reply packages/google_mlkit_document_scanner/android/src/main/java/com/google_mlkit_document_scanner/GoogleMlkitDocumentScannerPlugin.java:1]: Yeah! About that.
Separating the file won't work. 
I did that initially, but we need to have the FlutterPlugin and ActivityAware on the same abstract class. I spent sometime on it. thank for @ablause for bailing my out on it.

File: packages/google_mlkit_document_scanner/android/build.gradle (hunk 7 above)
Comment (bensonarafat) [reply packages/google_mlkit_document_scanner/android/build.gradle:29]: Okay

File: packages/google_mlkit_document_scanner/README.md (hunk 6 above)
Comment (bensonarafat) [reply packages/google_mlkit_document_scanner/README.md:53]: Yes, i added not support for iOS, but i am updating the requirement for iOS as you mentioned.

**Comment by bensonarafat** [comment]
> @bensonarafat : I have more comments.

Updated

File: packages/google_mlkit_document_scanner/README.md (hunk 6 above)
Comment (fbernaly) [reply packages/google_mlkit_document_scanner/README.md:53]: What I mean is: adding:

```
### iOS

Not supported
```

Remove what you added and replace with above.

File: packages/google_mlkit_document_scanner/ios/Classes/GoogleMlkitDocumentScannerPlugin.h [hunk 14]
```diff
@@ -0,0 +1,4 @@
+#import <Flutter/Flutter.h>
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/ios/Classes/GoogleMlkitDocumentScannerPlugin.h:1]: Name of this file should be: 
 `GoogleMlKitDocumentScannerPlugin.h`

File: packages/google_mlkit_document_scanner/ios/Classes/GoogleMlkitDocumentScannerPlugin.m [hunk 15]
```diff
@@ -0,0 +1,20 @@
+#import "GoogleMlKitDocumentScannerPlugin.h"
```
Comment (fbernaly) [inline packages/google_mlkit_document_scanner/ios/Classes/GoogleMlkitDocumentScannerPlugin.m:1]: Name of this file should be: 
 `GoogleMlKitDocumentScannerPlugin.m`

**Comment by fbernaly** [comment]
@bensonarafat : I added some more comments, and branch is out-of date, update branch.

File: packages/google_mlkit_document_scanner/ios/Classes/GoogleMlkitDocumentScannerPlugin.h (hunk 14 above)
Comment (bensonarafat) [reply packages/google_mlkit_document_scanner/ios/Classes/GoogleMlkitDocumentScannerPlugin.h:1]: okay, i thought i changed this.

**Comment by bensonarafat** [comment]
> @bensonarafat : I added some more comments, and branch is out-of date, update branch.

@fbernaly Branch updated and review updated

**Comment by fbernaly** [comment]
@bensonarafat : thanks for your contribution, I am merging this into feature/doc_scanner and play with it, once I confirm the feature is working as expected I will merge into master and cut a release.

**Pull request merged**